wgsl/index.bs

```diff
@@ -6981,13 +6979,310 @@ See [[#statements]] and [[#function-calls]].
+        One way to understand this, is that RequiresUniform corresponds to the proposition True, so that RequiresUniform -> X is the same as saying that X is true.
+
+        Reciprocally, to express that we cannot ensure the uniformity of something (e.g. a variable which holds the thread id), we simply add an edge from the corresponding node to AssumeNonUniform.
+        One way to understand this, is that AssumeNonUniform corresponds to the proposition False, so that X -> AssumeNonUniform is the same as saying that X is false.
+
+        A consequence of this interpretation is that every node reachable from RequiresUniform corresponds to something which must be uniform for the program to be valid, and every node from which cannotBeUniform is reachable corresponds to something whose uniformity we cannot guarantee. It follows that we have a uniformity violation (and thus reject the program) if there is any path from RequiresUniform to AssumeNonUniform.
+</div>
+
+In more detail, here is what we compute for each function *F*:
+
+* A <dfn>call-site-uniformity requirement</dfn>, a tag summarizing the control flow uniformity requirements at the function's call sites, one of:
+    * RequiresUniform: The function must be called when control flow is uniform.
+    * PropagatesFlowUniformity: The function does not induce a uniformity requirement at the call site,
+         but the call site may carry a uniformity requirement from code that executes after the call.
+* A <dfn>argument-uniformity requirement</dfn>, for each formal parameter to *F*, a tag summarizing the uniformity requirement on the
+       value of the corresponding argument at any call site to *F*, one of:
+    * RequiresUniform: The value must be uniform
+    * PropagatesFlowUniformity: The evaluation of the function does not require this argument to be uniform,
+        but the argument carries a control flow uniformity requirement from the code that executes after the call.
+        That is, if the argument might have a non-uniform value, then control flow after the call to F may be non-uniform.
+    * PropagatesValueUniformity: The evaluation of the function does not require this argument to be uniform,
+        but the argument carries a value uniformity requirement from the function return value.
+        That is, if the argument might have a non-uniform value, then the return value may be non-uniform.
+    * NoRestriction: The argument does not have or propagate any uniformity requirement.
+* A <dfn>call-site-uniformity effect</dfn>, a tag summarizing how calling this function affects subsequent computation, one of:
+    * AssumeNonUniformFlow: After calling this function, control flow is assumed to be non-uniform.
+    * AssumeNonUniformValue: The return value of this function is assumed to be non-uniform.
+    * NoRestriction: Calling this function has no effect on uniformity. (??)
```

Comment:
This is a large change so we can always defer to after this PR is landed. I could take on doing it too if you'd like.

Note: I found these ideas on how to rephrase the metadata when thinking how an author could explicitly annotate a function with the metadata:

```rust
[[uniform(workgroup)]] // call site uniformity, short hand because most common.
[[return_uniformity_min(subgroup, carrot)]] // Defaults to the call site uniformity if not specified?
fn f([[uniform(none)]] carrot: u32, [[uniform]] lettuce : u32) -> [[uniform(none)]] u32 {
  // ...
}
```